src/crimson/os/seastore/journal.cc

```diff
@@ -118,35 +119,66 @@ Journal::write_record_ret Journal::write_record(
     target);
 
   auto segment_id = current_journal_segment->get_segment_id();
+  submit_queue.emplace_back(std::make_tuple(target, current_journal_segment, std::move(to_write)));
 
-  // Start write under the current exclusive stage, but wait for it
-  // in the device_submission concurrent stage to permit multiple
-  // overlapping writes.
-  auto write_fut = current_journal_segment->write(target, to_write);
   return handle.enter(write_pipeline->device_submission
-  ).then([write_fut = std::move(write_fut)]() mutable {
-    return std::move(write_fut
-    ).handle_error(
-      write_record_ertr::pass_further{},
-      crimson::ct_error::assert_all{
-	"Invalid error in Journal::write_record"
-      }
-    );
-  }).safe_then([this, &handle] {
+  ).then([this, &handle, target, segment_id] {
     return handle.enter(write_pipeline->finalize);
-  }).safe_then([this, target, segment_id] {
+  }).then([this, target, segment_id]() -> write_record_ret {
     logger().debug(
       "write_record: commit target {}",
       target);
     if (segment_id == current_journal_segment->get_segment_id()) {
       assert(committed_to < target);
       committed_to = target;
     }
-    return write_record_ret(
-      write_record_ertr::ready_future_marker{},
-      paddr_t{
-	segment_id,
-	target});
+    if(!write_queue.empty() || submit_queue.empty()) {
+      return write_record_ret(
+        write_record_ertr::ready_future_marker{},
+        paddr_t{
+          segment_id,
+          target});
+    } else {
+      write_queue.swap(submit_queue);
+      logger().error("write_queue size = {}", write_queue.size());
+      return crimson::repeat([this] {
+        bufferlist bl;
+        auto tuple = write_queue.front();
+        auto [write_target, write_journal_segment, to_write] = tuple;
+        bl.append(to_write);
+        write_queue.pop_front();
+        while (!write_queue.empty()) {
+          auto tmp_tuple = write_queue.front();
+          auto [tmp_target, tmp_journal_segment, tmp_write] = tmp_tuple;
+          if (write_journal_segment == tmp_journal_segment) {
+            bl.append(tmp_write);
+            write_queue.pop_front();
+          } else {
+            break;
+          }
+        }
+        bl.rebuild();
+        return write_journal_segment->write(write_target, bl)
+          .safe_then([this] {
+            if (write_queue.empty()){
+              return seastar::stop_iteration::yes;
+            } else {
+              return seastar::stop_iteration::no;
+            }
+        });
+      }).safe_then ([this, target, segment_id] {
```

Comment:
We should be updating committed_to here, I think.

Reply:
committed_to is used in encode_record for each record convert to bufferlist,(record header use it)  every to_write bufferlist is push back to submit_queue, will assume writing finish. and update committed_to to for next coming record.  if put it here, the big bufferlist include several record, committed_to will be wrong in its header.

Reply:
committed_to should always refer to the latest write *that has been committed*.  When we perform a write, we write down the current committed_to, which would point to a prior completed write, and then update committed_to to point to the write just completed.